src/qopenglwebpage.cpp

```diff
@@ -0,0 +1,624 @@
+
+    \sa QOpenGLWebPage::requestGLContext()
+*/
+QOpenGLWebPage::QOpenGLWebPage(QObject *parent)
+  : QObject(parent)
+  , d(new QGraphicsMozViewPrivate(new IMozQView<QOpenGLWebPage>(*this), this))
+  , mParentID(0)
+  , mPrivateMode(false)
+  , mActive(true)
+  , mBackground(false)
+  , mLoaded(false)
+  , mCompleted(false)
+{
+    d->mContext = QMozContext::GetInstance();
+    d->mHasContext = true;
+
+    connect(this, SIGNAL(setIsActive(bool)), this, SLOT(SetIsActive(bool)));
+    connect(this, SIGNAL(viewInitialized()), this, SLOT(processViewInitialization()));
+    connect(this, SIGNAL(loadProgressChanged()), this, SLOT(updateLoaded()));
+    connect(this, SIGNAL(loadingChanged()), this, SLOT(updateLoaded()));
+
+    if (!d->mContext->initialized()) {
+        connect(d->mContext, SIGNAL(onInitialized()), this, SLOT(createView()));
+    } else {
+        createView();
+    }
+}
+
+QOpenGLWebPage::~QOpenGLWebPage()
+{
+    if (d->mView) {
+        d->mView->SetListener(NULL);
+        d->mContext->GetApp()->DestroyView(d->mView);
+    }
+    delete d;
+}
+
+void QOpenGLWebPage::SetIsActive(bool aIsActive)
+{
+    if (d->mView) {
+        d->mView->SetIsActive(aIsActive);
+        if (aIsActive) {
+            d->mView->ResumeRendering();
+        } else {
+            d->mView->SuspendRendering();
+        }
+    }
+}
+
+void QOpenGLWebPage::updateLoaded()
+{
+    bool loaded = loadProgress() == 100 && !loading();
+    if (mLoaded != loaded) {
+        mLoaded = loaded;
+        Q_EMIT loadedChanged();
+    }
+}
+
+void QOpenGLWebPage::createView()
+{
+    LOGT("QOpenGLWebPage");
+    if (!d->mView) {
+        // We really don't care about SW rendering on Qt5 anymore
+        d->mContext->GetApp()->SetIsAccelerated(true);
+        d->mView = d->mContext->GetApp()->CreateView(mParentID, mPrivateMode);
+        d->mView->SetListener(d);
+    }
+}
+
+void QOpenGLWebPage::processViewInitialization()
+{
+    // This is connected to view initialization. View must be initialized
+    // over here.
+    Q_ASSERT(d->mViewInitialized);
+
+    mCompleted = true;
+    forceActiveFocus();
+    Q_EMIT completedChanged();
+}
+
+void QOpenGLWebPage::createGeckoGLContext()
+{
+}
+
+/*!
+    \fn void QOpenGLWebPage::requestGLContext()
+
+    With in the slot connected to the requestGLContext() prepare QOpenGLContext and
+    make it current context of the gecko compositor thread,
+    against the surface of the QWindow.
+
+    \code
+    m_context = new QOpenGLContext();
+    m_context->setFormat(requestedFormat());
+    m_context->create();
+    m_context->makeCurrent(this);
+    \endcode
+
+    This signal is emitted from the gecko compositor thread, you must make sure that
+    the connection is direct (see Qt::ConnectionType)
+
+    \sa QOpenGLContext::makeCurrent(QSurface*)
+*/
+void QOpenGLWebPage::requestGLContext(bool& hasContext, QSize& viewPortSize)
+{
+    hasContext = true;
+    viewPortSize = d->mGLSurfaceSize;
+    Q_EMIT requestGLContext();
```

Comment:
Let's if it makes sense to have QMozView (QWindow) over here. One problematic case is settings the chrome window as transient parent needs to be set for the chrome window before it can be shown [1]. Stack order wise (z-order) browser content window is at background and chrome at foreground.

[1]
https://github.com/rainemak/sailfish-browser/commit/421831eab9eb1a463944cc54659fb7bb239555db#diff-95069946c6b81116a73994756f19ebfdR301

Here is a proto https://github.com/rainemak/qmlmozbrowser/blob/chromewindow/quick/browserwindow.cpp
and here a sailfish-browser wip branch: https://github.com/rainemak/sailfish-browser/commits/next